components/FilterableNumbersList.js

```diff
@@ -1,18 +1,47 @@
 import * as React from 'react';
 import { NumbersList } from './NumbersList';
+import { withRouter } from 'next/router';
+import { DebounceInput } from 'react-debounce-input';
+
+class _FilterableNumbersList extends React.Component {
+  state = {
+    query: this.props.router.query.q
+  };
+
+  updateRouter = query => {
+    this.props.router.replace({
+      pathname: '/',
+      query: { q: query }
+    });
+  };
+
+  updateQuery = e => {
+    const value = e.target.value;
+    this.setState({ query: value }, () => {
+      this.updateRouter(value);
+    });
+  };
 
-export class FilterableNumbersList extends React.Component {
   render() {
+    const { numbers } = this.props;
+    const searchInput = this.state.query;
+    const filteredNumbers = searchInput
+      ? numbers.filter(number => number.includes(searchInput))
+      : numbers;
+
     return (
       <section>
-        <input
+        <DebounceInput
           type="text"
           placeholder="Search numbers"
-          value={this.props.query}
-          onChange={this.props.setQuery}
+          value={this.state.query}
+          onChange={this.updateQuery}
+          debounceTimeout={300}
         />
-        <NumbersList numbers={this.props.numbers} />
+        <NumbersList numbers={filteredNumbers} />
       </section>
     );
   }
 }
+
+export const FilterableNumbersList = withRouter(_FilterableNumbersList);
```

Comment:
I debated whether to move the routing into this component vs keeping it in the parent and passing it in. Perhaps that would have been a better approach, as routing is a separate concern from what this component seems to be doing (filtering numbers and passing them off to be rendered). 

My thinking was to do this in a simpler way for now, but here is the approach I was considering:
- `pages/index.js` defines how to update the router
- It passes it to this component as an `onChange` prop
- We define a new `onChange` here which updates the query and calls the `onChange` prop